Source/Core/VideoCommon/BPMemory.h

```diff
@@ -544,109 +653,153 @@ struct FourTexUnits
-    u32 lineoff : 3;
-    u32 pointoff : 3;
-    u32 lineaspect : 1;  // interlacing: adjust for pixels having AR of 1/2
-    u32 padding : 1;
-  };
+  BitField<0, 8, u32> linesize;   // in 1/6th pixels
+  BitField<8, 8, u32> pointsize;  // in 1/6th pixels
+  BitField<16, 3, u32> lineoff;
+  BitField<19, 3, u32> pointoff;
+  // interlacing: adjust for pixels having AR of 1/2
+  BitField<22, 1, AspectRatioAdjustment> adjust_for_aspect_ratio;
   u32 hex;
 };
 
 union X12Y12
 {
-  struct
-  {
-    u32 y : 12;
-    u32 x : 12;
-  };
+  BitField<0, 12, u32> y;
+  BitField<12, 12, u32> x;
   u32 hex;
 };
 union X10Y10
 {
-  struct
-  {
-    u32 x : 10;
-    u32 y : 10;
-  };
+  BitField<0, 10, u32> x;
+  BitField<10, 10, u32> y;
   u32 hex;
 };
 
 // Framebuffer/pixel stuff (incl fog)
+enum class SrcBlendFactor : u32
+{
+  Zero = 0,
+  One = 1,
+  DstClr = 2,
+  InvDstClr = 3,
+  SrcAlpha = 4,
+  InvSrcAlpha = 5,
+  DstAlpha = 6,
+  InvDstAlpha = 7
+};
+template <>
+struct fmt::formatter<SrcBlendFactor> : EnumFormatter<SrcBlendFactor::InvDstAlpha>
+{
+  static constexpr array_type names = {"0",         "1",           "dst_color", "1-dst_color",
+                                       "src_alpha", "1-src_alpha", "dst_alpha", "1-dst_alpha"};
+  formatter() : EnumFormatter(names) {}
+};
 
-union BlendMode
+enum class DstBlendFactor : u32
 {
-  enum BlendFactor : u32
-  {
-    ZERO = 0,
-    ONE = 1,
-    SRCCLR = 2,             // for dst factor
-    INVSRCCLR = 3,          // for dst factor
-    DSTCLR = SRCCLR,        // for src factor
-    INVDSTCLR = INVSRCCLR,  // for src factor
-    SRCALPHA = 4,
-    INVSRCALPHA = 5,
-    DSTALPHA = 6,
-    INVDSTALPHA = 7
-  };
+  Zero = 0,
+  One = 1,
+  SrcClr = 2,
+  InvSrcClr = 3,
+  SrcAlpha = 4,
+  InvSrcAlpha = 5,
+  DstAlpha = 6,
+  InvDstAlpha = 7
+};
+template <>
+struct fmt::formatter<DstBlendFactor> : EnumFormatter<DstBlendFactor::InvDstAlpha>
+{
+  static constexpr array_type names = {"0",         "1",           "src_color", "1-src_color",
+                                       "src_alpha", "1-src_alpha", "dst_alpha", "1-dst_alpha"};
+  formatter() : EnumFormatter(names) {}
+};
 
-  enum LogicOp : u32
-  {
-    CLEAR = 0,
-    AND = 1,
-    AND_REVERSE = 2,
-    COPY = 3,
-    AND_INVERTED = 4,
-    NOOP = 5,
-    XOR = 6,
-    OR = 7,
-    NOR = 8,
-    EQUIV = 9,
-    INVERT = 10,
-    OR_REVERSE = 11,
-    COPY_INVERTED = 12,
-    OR_INVERTED = 13,
-    NAND = 14,
-    SET = 15
+enum class LogicOp : u32
+{
+  Clear = 0,
+  And = 1,
+  AndReverse = 2,
+  Copy = 3,
+  AndInverted = 4,
+  NoOp = 5,
+  Xor = 6,
+  Or = 7,
+  Nor = 8,
+  Equiv = 9,
+  Invert = 10,
+  OrReverse = 11,
+  CopyInverted = 12,
+  OrInverted = 13,
+  Nand = 14,
+  Set = 15
+};
+template <>
+struct fmt::formatter<LogicOp> : EnumFormatter<LogicOp::Set>
+{
+  static constexpr array_type names = {
+      "0",        "s & d",    "s & ~d", "s",      "~s & d", "d",      "s ^ d",    "s | d",
```

Comment:
This is nice but I'm wondering if the text in the enum would be more easily digested at a glance

Reply:
These names were [already present](https://github.com/dolphin-emu/dolphin/blob/9d0983c9c988f81c73c9a4ed61835ca9ac20db06/Source/Core/VideoCommon/BPStructs.cpp#L922-L924) before.  I agree that they're a bit hard to read at a glance; I could replace it with something like this:

```C++
  static constexpr array_type names = {
      "Clear (0)",
      "And (src & dst)",
      "And Reverse (src & ~dst)",
      "Copy (src)",
      "And Inverted (~src & dst)",
      "NoOp (dst)",
      "Xor (src ^ dst)",
      "Or (src | dst)",
      "Nor (~(src | dst))",
      "Equiv (~(src ^ dst))",
      "Invert (~dst)",
      "Or Reverse (src | ~dst)",
      "Copy Inverted (~src)",
      "Or Inverted (~src | dst)",
      "Nand (~(src & dst))",
      "Set (1)",
  };
```

Reply:
If you don't mind, I think that would be a good change.